### PR TITLE
Practice3 (not finished yet)

### PR DESCRIPTION
**Pull request by nick-krasnitskiy**
1. Add search history list.
2. Implement delete search history cell as a trailing swipe action.

### CODE REVIEW

File: DogsList/DogsList/Controllers/SearchHistoryViewController.swift [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            }
+        } else {
+            DispatchQueue.main.async {
+                self.dataSource.apply(snapshot, animatingDifferences: false)
+            }
+        }
+    }
+}
+
+extension Array where Element: Hashable {
```
Comment (alekseykorol) [inline DogsList/DogsList/Controllers/SearchHistoryViewController.swift]: Move extension to separate file

File: DogsList/DogsList/Controllers/SearchHistoryViewController.swift [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        configureDataSource()
+        addSnaphot(searchHistory: searchHistory.removingDuplicates())
+        navigationItem.title = "Search History"
+        navigationController?.navigationBar.tintColor = .white
+    }
+}
+
+extension SearchHistoryViewController {
+    
+    private func createLayout() -> UICollectionViewLayout {
```
Comment (alekseykorol) [inline DogsList/DogsList/Controllers/SearchHistoryViewController.swift:43]: You can group all private methods under a private extension. 

```
private extension SearchHistoryViewController {
    func createLayout() -> UICollectionViewLayout {
    ...
    }
    ...
}
```

File: DogsList/DogsList/Controllers/Cells/StarWarsCells/StarWarsSearchHistoryViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+//  DogsList
+//
+//  Created by Nick Krasnitskiy on 16.07.2021.
+//
+
+import UIKit
+
+class StarWarsSearchHistoryViewCell: UICollectionViewCell {
+
+    @IBOutlet weak var query: UILabel!
```
Comment (alekseykorol) [inline DogsList/DogsList/Controllers/Cells/StarWarsCells/StarWarsSearchHistoryViewCell.swift]: IBOutlets should be private

File: DogsList/DogsList/Controllers/SearchHistoryViewController.swift [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        self.searchHistory = searchHistory
+        super.init(nibName: nil, bundle: nil)
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    typealias DataSource = UICollectionViewDiffableDataSource<Section, String>
+    private var dataSource: DataSource!
```
Comment (vasilevkin) [inline DogsList/DogsList/Controllers/SearchHistoryViewController.swift:28]: do not use force unwrap for optionals

File: DogsList/DogsList/Controllers/SearchHistoryViewController.swift [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    private var collectionView: UICollectionView! = nil
```
Comment (vasilevkin) [inline DogsList/DogsList/Controllers/SearchHistoryViewController.swift:29]: = nil , is not required

File: DogsList/DogsList/Controllers/SearchStarWarViewController.swift [hunk 6]
```diff
@@ -67,7 +75,7 @@ extension SearchStarWarViewController {
     private func alertGeenrate(alertTitle: String, alertMessage: String, actionTitle: String, handler: ((UIAlertAction) -> Void)? = nil) {
         let alert = UIAlertController(title: alertTitle, message: alertMessage, preferredStyle: .alert)
         alert.addAction(UIAlertAction(title: actionTitle, style: .default) {(_: UIAlertAction!) in
-            self.viewDidLoad()})
+                            self.viewDidLoad()})
```
Comment (vasilevkin) [inline DogsList/DogsList/Controllers/SearchStarWarViewController.swift:78]: add SwiftLint to the project and fix all errors/warnings

**Review by vasilevkin — CHANGES_REQUESTED** [review]

File: DogsList/DogsList/Controllers/SearchHistoryViewController.swift (hunk 1 above)
Comment (nick-krasnitskiy) [reply DogsList/DogsList/Controllers/SearchHistoryViewController.swift]: Ok. Fixed.

File: DogsList/DogsList/Controllers/SearchHistoryViewController.swift (hunk 2 above)
Comment (nick-krasnitskiy) [reply DogsList/DogsList/Controllers/SearchHistoryViewController.swift:43]: Ok. Fixed.

File: DogsList/DogsList/Controllers/Cells/StarWarsCells/StarWarsSearchHistoryViewCell.swift (hunk 3 above)
Comment (nick-krasnitskiy) [reply DogsList/DogsList/Controllers/Cells/StarWarsCells/StarWarsSearchHistoryViewCell.swift]: Ok. Fixed.

File: DogsList/DogsList/Controllers/SearchHistoryViewController.swift (hunk 4 above)
Comment (nick-krasnitskiy) [reply DogsList/DogsList/Controllers/SearchHistoryViewController.swift:28]: I thought it's an Apple standard practice from its samples.

File: DogsList/DogsList/Controllers/SearchHistoryViewController.swift (hunk 5 above)
Comment (nick-krasnitskiy) [reply DogsList/DogsList/Controllers/SearchHistoryViewController.swift:29]: I thought it's an Apple standard practice from its samples.

File: DogsList/DogsList/Controllers/SearchStarWarViewController.swift (hunk 6 above)
Comment (nick-krasnitskiy) [reply DogsList/DogsList/Controllers/SearchStarWarViewController.swift:78]: Ok. SwiftLint has already been added.

File: DogsList/DogsList/Controllers/SearchHistoryViewController.swift (hunk 4 above)
Comment (vasilevkin) [reply DogsList/DogsList/Controllers/SearchHistoryViewController.swift:28]: no, it's used in samples for brevity only.

**Review by vasilevkin — APPROVED** [review]
LGTM

**Pull request merged**